File: android/src/main/kotlin/com/openland/react/animations/RNOAnimationSpec.kt
package com.openland.react.animations

import android.util.Log
import com.beust.klaxon.JsonArray
import com.beust.klaxon.JsonObject
import com.beust.klaxon.Parser

class RNOAnimationTransactionSpec {
    var transactionKey: String? = null
    var duration: Float = 0.3f
    var animations = mutableListOf<RNOAnimationSpec>()
    var valueSets = mutableListOf<RNOValueSet>()
}

enum class RNOAnimationType {
    timing,
    circular
}

enum class RNOEasingType {
    material,
    linear,
    bezier
}

class RNOEasing {
    var type: RNOEasingType = RNOEasingType.linear
    var bezier: Array<Float>? = null
}

class RNOAnimationSpec {
    var type: RNOAnimationType = RNOAnimationType.timing
    var easing: RNOEasing = RNOEasing()
    var viewKey: String = ""
    var property: String = ""
    var to: Float = 0.0f
    var from: Float = 0.0f

    var centerX: Float = 0.0f
    var centerY: Float = 0.0f

    var duration: Float? = null
    var optional: Boolean = false
}

class RNOValueSet {
    var viewKey: String = ""
    var property: String = ""
    var value: Float = 0.0f
    var optional: Boolean = false
}

fun parseAnimationSpec(spec: String): RNOAnimationTransactionSpec {
    val res = RNOAnimationTransactionSpec()
    val parser = Parser()
    val parsed = parser.parse(StringBuilder(spec)) as JsonObject
    if (parsed["duration"] is Number) {
        res.duration = (parsed["duration"] as Number).toFloat()
    }
    if (parsed["transactionKey"] is String) {
        res.transactionKey = parsed["transactionKey"] as String
    }
    for (anim in parsed["animations"] as JsonArray<JsonObject>) {
        val aspec = RNOAnimationSpec()
        when (anim["type"] as String) {
            "spring" -> {
                Log.d("RNOAnimated", "Spring animations are not supported on Android")
            }
            "timing" -> aspec.type = RNOAnimationType.timing
            "circular" -> {
                aspec.type = RNOAnimationType.circular
            }
        }

        aspec.viewKey = anim["view"] as String
        aspec.to = (anim["to"] as Number).toFloat()
        aspec.from = (anim["from"] as Number).toFloat()

        // Property
        if (aspec.type != RNOAnimationType.circular) {
            aspec.property = anim["prop"] as String
        }

        // Easing
        if (anim["easing"] is JsonObject) {
            val eas = anim["easing"] as JsonObject
            when(eas["type"] as String) {
                "linear" -> aspec.easing = RNOEasing().apply { type = RNOEasingType.linear }
                "material" -> aspec.easing = RNOEasing().apply { type = RNOEasingType.material }
                "bezier" -> aspec.easing = RNOEasing().apply { type = RNOEasingType.bezier; bezier = (eas["bezier"] as JsonArray<Number>).map { it.toFloat() }.toTypedArray() }
            }
        }

        // Duration
        if (anim["duration"] is Number) {
            aspec.duration = (anim["duration"] as Number).toFloat()
        }

        if (anim["centerX"] is Number) {
            aspec.centerX = (anim["centerX"] as Number).toFloat()
        }
        if (anim["centerY"] is Number) {
            aspec.centerY = (anim["centerY"] as Number).toFloat()
        }

        // Optional
        if (anim["optional"] is Boolean) {
            aspec.optional = anim["optional"] as Boolean
        }

        res.animations.add(aspec)
    }
    for (setter in parsed["valueSetters"] as JsonArray<JsonObject>) {
        val aspec = RNOValueSet()
        aspec.viewKey = setter["view"] as String
        aspec.property = setter["prop"] as String
        aspec.value = (setter["to"] as Number).toFloat()
        if (setter["optional"] is Boolean) {
            aspec.optional = setter["optional"] as Boolean
        }
        res.valueSets.add(aspec)
    }
    return res
}

File: ios/Sources/RNOAnimationSpec.swift
//
//  RNFastAnimationSpec.swift
//  openland
//
//  Created by Steve Kite on 9/10/18.
//  Copyright © 2018 Openland. All rights reserved.
//

import Foundation
import SwiftyJSON

class RNOAnimationTransactionSpec {
  
  static func parse(spec: String) -> RNOAnimationTransactionSpec {
    let src = JSON(parseJSON: spec)
    let res = RNOAnimationTransactionSpec()
    if let transactionKey = src["transactionKey"].string {
      res.transactionKey = transactionKey
    }
    if let duration = src["duration"].double {
      res.duration = duration
    }
    if let animations = src["animations"].array {
      res.animations = animations.map { (anim) -> RNOAnimationSpec in
        let aspec = RNOAnimationSpec()
      
        // Type and View
        aspec.type = RNOAnimationType(rawValue: anim["type"].stringValue)!
        aspec.viewKey = anim["view"].string!
        aspec.property = anim["prop"].string!
        aspec.to = CGFloat(anim["to"].double!)
        aspec.from = CGFloat(anim["from"].double!)
        
        // Duration
        if let duration = anim["duration"].double {
          aspec.duration = duration
        }
        
        if let delay = anim["delay"].double {
          aspec.delay = delay
        }
        
        if let velocity = anim["velocity"].double {
          aspec.velocity = CGFloat(velocity)
        }
        
        // Can we ignore this animation if view is missing?
        if let optional = anim["optional"].bool {
          aspec.optional = optional
        }
        
        return aspec
      }
    }
    if let valueSets = src["valueSetters"].array {
      res.valueSets = valueSets.map({ (s) -> RNOValueSetSpec in
        let aspec = RNOValueSetSpec()
        aspec.viewKey = s["view"].string!
        aspec.property = s["prop"].string!
        aspec.value = CGFloat(s["to"].double!)
        
        // Can we ignore this animation if view is missing?
        if let optional = s["optional"].bool {
          aspec.optional = optional
        }
        
        return aspec
      })
    }
    return res
  }
  
  var transactionKey: String?
  var animations: [RNOAnimationSpec] = []
  var valueSets: [RNOValueSetSpec] = []
  var duration: Double = 0.3
}

enum RNOAnimationType: String {
  case spring = "spring"
  case timing = "timing"
}

class RNOValueSetSpec {
  var viewKey: String!
  var property: String!
  var value: CGFloat!
  var optional: Bool = false
}

class RNOAnimationSpec {
  var type: RNOAnimationType!
  var viewKey: String!
  var property: String!
  var to: CGFloat!
  var from: CGFloat!
  var velocity: CGFloat?
  
  var duration: Double?
  var delay: Double?
  var optional: Bool = false
}
